File: services/hardware-helper/Models/HelperMessage.cs
using System.Text.Json.Serialization;

namespace HardwareHelper.Models;

public sealed class HelperMessage
{
    [JsonPropertyName("type")]
    public string Type { get; set; } = string.Empty;

    [JsonPropertyName("version")]
    public string Version { get; set; } = "1.0.0";

    [JsonPropertyName("timestamp")]
    public long Timestamp { get; set; }

    [JsonPropertyName("data")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HardwareSnapshot? Data { get; set; }

    [JsonPropertyName("message")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Message { get; set; }

    [JsonPropertyName("recoverable")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? Recoverable { get; set; }

    public static HelperMessage CreateInit()
    {
        return new HelperMessage
        {
            Type = "init",
            Timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
        };
    }

    public static HelperMessage CreateSnapshot(HardwareSnapshot data)
    {
        return new HelperMessage
        {
            Type = "snapshot",
            Timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
            Data = data,
        };
    }

    public static HelperMessage CreateError(string message, bool recoverable)
    {
        return new HelperMessage
        {
            Type = "error",
            Timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
            Message = message,
            Recoverable = recoverable,
        };
    }

    public static HelperMessage CreateStatus(string message)
    {
        return new HelperMessage
        {
            Type = "status",
            Timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
            Message = message,
        };
    }
}

File: packages/shared-types/helper_message.ts
// packages/shared-types/helper_message.ts
// Contract version: 1.0.0
// Source schema: packages/contracts/helper_message.schema.json
//
// Envelope protocol for all stdout output from hardware-helper.exe.
// Every line is a complete HelperMessage JSON object.

import type { HardwareSnapshot } from './hardware';

export type HelperMessageType = 'init' | 'snapshot' | 'error' | 'status';

export interface HelperMessageBase {
  type: HelperMessageType;
  version: string;
  timestamp: number;
}

export interface HelperMessageInit extends HelperMessageBase {
  type: 'init';
}

export interface HelperMessageSnapshot extends HelperMessageBase {
  type: 'snapshot';
  data: HardwareSnapshot;
}

export interface HelperMessageError extends HelperMessageBase {
  type: 'error';
  message: string;
  recoverable: boolean;
}

export interface HelperMessageStatus extends HelperMessageBase {
  type: 'status';
  message: string;
}

export type HelperMessage =
  | HelperMessageInit
  | HelperMessageSnapshot
  | HelperMessageError
  | HelperMessageStatus;

/**
 * Type guard: checks if a HelperMessage is a snapshot message.
 * Useful for filtering snapshot lines from the stdout stream.
 */
export function isSnapshotMessage(msg: HelperMessage): msg is HelperMessageSnapshot {
  return msg.type === 'snapshot';
}

/**
 * Type guard: checks if an error is recoverable.
 * Recoverable errors: keep collecting. Non-recoverable: helper is about to exit.
 */
export function isRecoverableError(msg: HelperMessage): msg is HelperMessageError {
  return msg.type === 'error' && msg.recoverable === true;
}
